placement/algo/sharded_helper.go

```diff
@@ -738,3 +738,80 @@ func newShards(shardIDs []uint32) []shard.Shard {
 	}
 	return r
 }
+
+func removeInstanceFromList(list []placement.Instance, instanceID string) []placement.Instance {
+	for i, instance := range list {
+		if instance.ID() == instanceID {
+			last := len(list) - 1
+			list[i] = list[last]
+			return list[:last]
+		}
+	}
+	return list
+}
+
+func markShardAvailable(p placement.Placement, instanceID string, shardID uint32, nowNanos int64, shouldCheckCutoverTime bool) (placement.Placement, error) {
+	instance, exist := p.Instance(instanceID)
+	if !exist {
+		return nil, fmt.Errorf("instance %s does not exist in placement", instanceID)
+	}
+
+	shards := instance.Shards()
+	s, exist := shards.Shard(shardID)
+	if !exist {
+		return nil, fmt.Errorf("shard %d does not exist in instance %s", shardID, instanceID)
+	}
+
+	if s.State() != shard.Initializing {
+		return nil, fmt.Errorf("could not mark shard %d as available, it's not in Initializing state", s.ID())
+	}
+
+	if shouldCheckCutoverTime && s.CutoverNanos() > nowNanos {
+		return nil, fmt.Errorf("could not mark shard %d as available, it's cutover time %d is scheduled after now %d", s.ID(), s.CutoverNanos(), nowNanos)
+	}
+
+	sourceID := s.SourceID()
+	shards.Add(shard.NewShard(shardID).SetState(shard.Available))
+
+	// There could be no source for cases like initial placement.
+	if sourceID == "" {
+		return p, nil
+	}
+
+	sourceInstance, exist := p.Instance(sourceID)
+	if !exist {
+		return nil, fmt.Errorf("source instance %s for shard %d does not exist in placement", sourceID, shardID)
+	}
+
+	sourceShards := sourceInstance.Shards()
+	sourceShard, exist := sourceShards.Shard(shardID)
+	if !exist {
+		return nil, fmt.Errorf("shard %d does not exist in source instance %s", shardID, sourceID)
+	}
+
+	if sourceShard.State() != shard.Leaving {
+		return nil, fmt.Errorf("shard %d is not leaving instance %s", shardID, sourceID)
+	}
+
+	sourceShards.Remove(shardID)
```

Comment:
Do we need to consider cutoff times here? In particular, if a shard is cut off at 2pm, but doesn't flush all its data until 2:50pm, we probably don't want to remove the leaving shard from the placement, which causes the shard to be closed immediately on the instance. 

Perhaps we should set `shardCutoverFn` and `shardCutoffFn` on the placement options, and use the two functions to evaluate when a shard can be marked as available / removed, where the `shardCutoverFn` compares the cutover time against now, and the `shardCutoffFn` compares the cutoff + maxAggregationWindow against now.

Reply:
yeah good point, could compare the cutoff time + max window as well

Reply:
Probably don't want to introduce the concept of "max window" here though, which is specific to the agg tier placement generation. I'd just pass in the two functions and provide them in the tool for generating agg tier placement.

Reply:
? I just passed the 2 validation functions to the function? the m3admin tooling will be creating the cutoff validation function with max window taken into consideration

Reply:
Right yeah that's what I meant, not sure whether you were doing that so figured I should raise it, sounds good then.